Multiple buffer support and standardised window and buffer handling

mickeynp opened the pull request:
Hi Dirk-Jan,

Phew. OK, so, this (sadly rather large) PR fixes a large number of long-standing issues and standardises buffer display so it uses Emacs's `display-buffer`, `pop-to-buffer`, and action system so users can override everything in `display-buffer-alist`.

Here's what it does:

1. Separate (all, I think!) window and buffer management into `mu4e-window.el`;
2. Centralise the creation and retrieval of headers and view buffer into two functions;
3. Centralise the display of buffers to `mu4e-display-buffer`;
4. Rewrite all split window code to use `display-buffer` and the action display system;
5. Support multiple, distinct and standalone headers and views. Views are now detachable (`z` key) from a header, meaning it is inviolable and not touched by movement and navigation elsewhere.
6. Re-attachment (`Z`) to another headers buffer.

Things left to do, pending your approval:

1. Documentation update
2. Testing and bug fixing.

I use it, and it works well, though.

In practice this means you can write a `display-buffer-alist` rule to force buffers to display in tabs; in sidebars; in new frames; etc. with a simple rule.

Feedback welcome. Sorry about the massive PR, but it took quite a lot of effort to rip out all the window and buffer code from all the different places and centralise it.

mickeynp commented:
Note I recommend you enable ignore whitespace as I ended up indenting things with `C-M-\` that Emacs felt needed indenting.

djcb commented:
I played a bit with it...  it seems to work fine with some light testing... in particular `z` works nicely, and users have been asking about that feature! Cool.

Needed
 ```
modified   mu4e/meson.build
@@ -52,7 +52,8 @@ mu4e_srcs=[
   'mu4e-speedbar.el',
   'mu4e-update.el',
   'mu4e-vars.el',
-  'mu4e-view.el'
+  'mu4e-view.el',
+  'mu4e-window.el'
 ]
```
to work for my setup.

djcb commented:
There are a bunch of byte-compiler warnings:
```
In end of data:
../mu4e/mu4e-helpers.el:117:14: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-helpers.el:115:14: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-helpers.el:114:14: Warning: the function ‘mu4e-current-buffer-type-p’ is not known to be defined.
[29/96] Generating mu4e/mu4e_draft_el_el with a custom command

In mu4e~draft-open-file:
../mu4e/mu4e-draft.el:668:35: Warning: Unused lexical argument `switch-function'
[30/96] Generating mu4e/mu4e_compose_el_el with a custom command

In end of data:
../mu4e/mu4e-compose.el:613:25: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-compose.el:611:23: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-compose.el:607:6: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
[33/96] Generating mu4e/mu4e_main_el_el with a custom command

In mu4e--main-redraw-buffer:
../mu4e/mu4e-main.el:276:24: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-view:
../mu4e/mu4e-main.el:373:33: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-toggle-mail-sending-mode:
../mu4e/mu4e-main.el:395:47: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-update-after-index:
../mu4e/mu4e-main.el:419:24: Warning: reference to free variable ‘mu4e-main-buffer-name’

In end of data:
../mu4e/mu4e-main.el:377:6: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
[35/96] Generating mu4e/mu4e_mark_el_el with a custom command

In mu4e--mark-find-headers-buffer:
../mu4e/mu4e-mark.el:101:22: Warning: Unused lexical argument `b'

In end of data:
../mu4e/mu4e-mark.el: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-mark.el:102:16: Warning: the function ‘mu4e-current-buffer-type-p’ is not known to be defined.
[36/96] Generating mu4e/mu4e_headers_el_el with a custom command

In mu4e~headers-view-this-message-p:
../mu4e/mu4e-headers.el:727:35: Warning: Unused lexical argument `buf'

In mu4e~loading-close:
../mu4e/mu4e-headers.el:1788:15: Warning: reference to free variable ‘mu4e~headers-loading-buf’

In end of data:
../mu4e/mu4e-headers.el:843:8: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
../mu4e/mu4e-headers.el:808:30: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-headers.el:727:4: Warning: the function ‘mu4e-get-view-buffers’ is not known to be defined.
../mu4e/mu4e-headers.el:354:25: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
[38/96] Generating mu4e/mu4e_message_el_el with a custom command

In mu4e-sexp-at-point:
../mu4e/mu4e-message.el:212:36: Warning: reference to free variable ‘mu4e--sexp-buffer-name’
[41/96] Generating mu4e/mu4e_update_el_el with a custom command

In mu4e--temp-window:
../mu4e/mu4e-update.el:225:10: Warning: Unused lexical variable `win'
[44/96] Generating mu4e/mu4e_org_el_el with a custom command

In end of data:
../mu4e/mu4e-icalendar.el:194:10: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
../mu4e/mu4e-icalendar.el:193:34: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-icalendar.el:189:29: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
[51/96] Generating mu4e/mu4e_view_el_el with a custom command

In mu4e-view-detach:
../mu4e/mu4e-view.el:321:11: Warning: reference to free variable ‘mu4e-linked-headers-buffer’
../mu4e/mu4e-view.el:328:25: Warning: assignment to free variable ‘mu4e-linked-headers-buffer’

In mu4e-view-attach:
../mu4e/mu4e-view.el:338:9: Warning: assignment to free variable ‘mu4e-linked-headers-buffer’

In mu4e-view:
../mu4e/mu4e-view.el:624:26: Warning: reference to free variable ‘mu4e-linked-headers-buffer’

In end of data:
../mu4e/mu4e-view.el:611:32: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-view.el: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-view.el: Warning: the function ‘mu4e--view-detached-p’ is not known to be defined.
../mu4e/mu4e-view.el: Warning: the function ‘mu4e-current-buffer-type-p’ is not known to be defined.
../mu4e/mu4e-view.el:142:6: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
```

djcb commented:
Anyway, I'll test some more :)

mickeynp commented:
Yeah I've got some work to do to get everything looking nice, including docs and some other fixes. There's still the matter of mu4e' new frame support that I need to add in.

Also I had to sacrifice a few things:

1. The automatic resizing of the headers window when a view is popped up.
2. The loading mode buffer as it interferes with buffer ordering. I think a better approach is making it a minor mode and instead overlaying "Loading" on top. If there's a way of muting the text colours when it's active then that would look nice too. Need to investigate that.

djcb commented:
> Yeah I've got some work to do to get everything looking nice, including docs and some other fixes. There's still the matter of mu4e' new frame support that I need to add in.
> 
> Also I had to sacrifice a few things:
> 
>     1. The automatic resizing of the headers window when a view is popped up.

Yes, the view window changes its size as a flip through messages in the headers view; and the initial main screen is only half-screen now. But I guess that's only a temporary sacrifice?

>     2. The loading mode buffer as it interferes with buffer ordering. I think a better approach is making it a minor mode and instead overlaying "Loading" on top. If there's a way of muting the text colours when it's active then that would look nice too. Need to investigate that.

Ah yeah, that sounds nice!

djcb commented:
Nice work... yeah, walking through messages works nicely now. 

There's still the smaller issue of the main buffer not being fullscreen...  but maybe that's an improvement? Not sure, perhaps we can leave that as is and see what users says?

There are still a few compiler warnings (when I compile with emacs 29); probably need some `declare-function` [ see below]. Otherwise, is there anything still needing work? Or are you ready for merging?


```
In end of data:
../mu4e/mu4e-helpers.el:117:14: Warning: the function ‘mu4e-get-view-buffer’ is not known to be defined.
../mu4e/mu4e-helpers.el:115:14: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-helpers.el:114:14: Warning: the function ‘mu4e-current-buffer-type-p’ is not known to be defined.
[94/133] Generating mu4e/mu4e_message_el_el with a custom command

In mu4e-sexp-at-point:
../mu4e/mu4e-message.el:212:36: Warning: reference to free variable ‘mu4e--sexp-buffer-name’
[96/133] Generating mu4e/mu4e_mark_el_el with a custom command

In end of data:
../mu4e/mu4e-mark.el: Warning: the function ‘mu4e-get-headers-buffer’ is not known to be defined.
../mu4e/mu4e-mark.el:102:16: Warning: the function ‘mu4e-current-buffer-type-p’ is not known to be defined.
[98/133] Generating mu4e/mu4e_main_el_el with a custom command

In mu4e--main-redraw-buffer:
../mu4e/mu4e-main.el:276:24: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-view:
../mu4e/mu4e-main.el:373:33: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-toggle-mail-sending-mode:
../mu4e/mu4e-main.el:395:47: Warning: reference to free variable ‘mu4e-main-buffer-name’

In mu4e--main-update-after-index:
../mu4e/mu4e-main.el:419:24: Warning: reference to free variable ‘mu4e-main-buffer-name’

In end of data:
../mu4e/mu4e-main.el:377:6: Warning: the function ‘mu4e-display-buffer’ is not known to be defined.
[106/133] Generating mu4e/mu4e_window_el_el with a custom command

In mu4e-display-buffer:
../mu4e/mu4e-window.el:201:18: Warning: reference to free variable ‘mu4e-split-view’
../mu4e/mu4e-window.el:223:60: Warning: reference to free variable ‘mu4e-compose-in-new-frame’

In mu4e-resize-linked-headers-window:
../mu4e/mu4e-window.el:266:18: Warning: reference to free variable ‘mu4e-split-view’
../mu4e/mu4e-window.el:264:36: Warning: reference to free variable ‘mu4e-headers-visible-columns’
../mu4e/mu4e-window.el:267:42: Warning: reference to free variable ‘mu4e-headers-visible-lines’

In end of data:
../mu4e/mu4e-window.el:177:14: Warning: the function ‘mu4e-view-mode’ is not known to be defined.
[116/133] Generating mu4e/mu4e_contrib_el_el with a custom command
```

mickeynp commented:
> There's still the smaller issue of the main buffer not being fullscreen... but maybe that's an improvement? Not sure, perhaps we can leave that as is and see what users says?

I'm happy to fix that (I assume you mean `*mu4e-main*`) by changing `pop-to-buffer` to `pop-to-buffer-same-window` in `mu4e-display-buffer` but maybe we can see what people say? I didn't even notice the regression.

Had another go at fixing the byte compiler warnings: I moved some stuff around to try and lessen the need for `declare-function` but, I won't lie, the defcustom stuff --- where it should really go when it straddles so many parts of mu4e is not so easy for me to pick a logical place for it. 

Aside from that, I think it's at a point where it's not a direct *downgrade* to the current experience though, I do admit, it's a little bit different in places. But I think that can be tweaked as people start experimenting.

djcb commented:
Yeah, the warnings are gone now, good! And the new way for loading a message is a nice touch.
Anyway, I'll merge it now, thank your for work!

mickeynp commented:
No problem. Ping me in any issues if there are problems!

djcb commented:
Okay! @mickeynp , seems that in `mu4e-display-buffer`, when `mu4e-split-view` is a function it is used for _all_ buffert-types, not just the view/headers case. Or?

mickeynp commented:
That's right. Which is how I believe it was also implemented -- sort of -- before. There were many checks for `mu4e-split-view` in the code base, and with varying levels of complex handling depending on its value. 

If a function is set, it is given the proposed buffer name and must return a live window to `set-window-buffer`.

theophilusx commented:
Just as another data point, I'm observing some undesired behaviour
following the merge of these changes. Note that I've not made any
configuration changes to my setup and I don't have any custom functions
etc - just basic out of the box configuration.

I'm running Emacs 29 (built from git repository) on Fedora 37. I run
Emacs as a daemon, started via systemctl --user and have this script to
get integration with my wm (xfce)

#!/bin/bash

emacsclient -c -e "(mu4e)"

1. Previously, when I ran this script, a frame would open with the main
mu4e window/buffer in it. Now, the frame is split into two windows with
the scratch buffer in the top window and mu4e in the bottom. I have to
do an C-x 1 to get a single window.

2. Previously, when I was in a header view and I decided to reply to a
message, a new composition window would appear in the lower half of the
frame. I would compose my reply and on sending, saving to draft or
cancelling/killing, the compose window would close and I would be back
in the header view. Now, when I finish composing my message, I end up
with another header view window i.e. frame is split in two with two
header view windows instead of 1. I have to do C-x 1 to get a single
header window again. 

3. Previously, if I viewed a message from the header view and switched
to the message window I could move to previous/next message with no
problems. Now I frequently get an error and have to leave the message
window, returning to the header window and then do next/previous. The
error is usually about the message not existing (I'll try to note it
down next time it occurs).

I guess many of these issues could be fixed if I was to customize
display functions. However, I don't believe this should be
necessary. The behaviour I want (and had) seems like the right default
to me. i.e. runing M-x mu4e starts a full frame window with mu4e summary
window. On sending/saving/cancelling a message, the previous window
configuration is restored (either two windows - header and message
preview or single window with header view - basically whatever it was
prior to hitting compose/reply/forward etc. Finally, when in the message
view window, moving to the previous/next message from within that window
should work consistently. 

None of these issues are critical. I think the PRs from Mickey are a
positive addition, but a bit more work is required to get correct and
consistent default behaviour.

djcb commented:
The non-fullframe 'main' screen is easy to fix.  

But the other problem is a bit trickier.; summarizing a bit:

With default settings, e.g.
- case 1: - (1)go to inbox ->  (2)see headers -> (3)open message -> window splits -> (4)read message ->  (5)'q' . Window closes and we're back to (2). All good!
 
The more complicated step is
- case 2:   (4) read message -> (5) reply to message -> (6) done replying -> back to (4)
*but* for some reason when pressing 'q' now, the window is _not_ closed, but replaced by the headers window (so we have 2 now. [ I think it should behave _just_ like in case 1 ]

djcb commented:
Note I've pushed an update to go back to the previous full-frame main-screen; after some testing I think I prefer that behavior.

djcb commented:
@mickeynp: perhaps an example of using `display-buffer-alist` would be useful for the manual? It may be necessary to retire the functionp `mu4e-split-window`, but at least would be  nice to give users a hint how they can tweak it.

Also, any idea how we could restore the 'case 1'-behavior for case 2? Thanks!

djcb commented:
I'm experimenting with:
```
modified   mu4e/mu4e-view.el
@@ -126,8 +126,8 @@ specified a function as viewer."
 This is a rather complex function, to ensure we don't disturb
 other windows."
   (interactive)
-  (quit-window t))
-
+  (kill-buffer)
+  (delete-window))
```

mickeynp commented:
I've pushed a branch with some fixes for these issues. Can you please try it with `bugfix/fix-window-issues`?

You're right, you should not have to customize anything to get this behaviour: it ought to work by default. 

@djcb, can you try my branch without the patch you just made in `7d9a9d8a7c390f1116c38f8638d2479740f83a7c`?

djcb commented:
@mickeynp: I tried that branch, but I'm still still seeing the "case-2" problem. 

I've pushed a tiny fix (and some whitespace noise) that does the `kill-buffer-and-window`, which seems to work for that cases, not just for composing messages, but also with 'raw view' and when showing attachments in emacs.

Not sure if that's the ultimate solution, but it seems to work.

mickeynp commented:
Sounds good. The only problem with `quit-window' is if its buffer changes; at that point quit will no longer, well, quit the window, but keep it. That's default (by perhaps confusing) Emacs behaviour though: I suspect that is what's happening here.

Change seems good 👍

bjcohen commented:
@mickeynp I'm a little bit confused by the intent of some of your changes - `single-window` seems to be fairly broken and it's not obvious to me how `display-buffer-alist` can help.

Examples:

- Split frame with two windows, scroll to next via `mu4e-view-headers-next` -> layout is replaced by a single window with the next message in view.
- Mark/scroll via `mu4e-headers-mark-for-*`or `mu4e-headers-next` open -> mark is (sometimes) recorded and the next message is opened in view. Sometimes the mark is not recorded before the next message is opened.

It looks like your intent was to remove some special casing on `mu4e-split-view` values (eg, https://github.com/djcb/mu/commit/69a17bfcb93b4ca8ee59c546f14a3e921202690c#diff-06a824e134b290af28c381d257da143ccfc6c4d64493108eb7b5d7f62fb845d9L1694) - is there something in `mu4e-window` that is supposed to replace this check, and is it broken? Or is `single-window` no longer supported?

mickeynp commented:
I am little bit confused. What is the problem exactly? I changed my  `mu4e-split-view` to `single-window` and... it does what it is supposed to do:

1. Split frame into a random grab bag of windows -- none of which are mu4e.
2. `M-x mu4e`. It opens up in the selected window. This is correct.
3. I open a random bookmark. It supplants the `mu4e-main` window, which is also the same as the selected window. That also seems right.
4. I open a message. Same thing as point 3; the selected window is used.
5. I navigate up or down. Everything recycles the same window.

At no point does mu4e trash your window layout. The one exception to this is when you quit stuff. Dirk-Jan prefers an approach that kills the buffer and window to `quit-window`, which was my original implementation, which has a quirk/feature/gotcha where if you switch buffers in a window that was displayed by `display-buffer`, then `quit-window` will instead restore the previously-seen buffer. (D.J. preferred that it instead just nuke the mu4e window, which is fine.)

djcb commented:
@bjcohen: ah, those are the steps to re-create the problem! there was some discussion in #2385, but your steps let me finally reproduce it.  It's not meant to work like that.

The problem case is when you do _not_ open a message but just walk through the headers; `n` then opens a message when it shouldn't.

bjcohen commented:
@djcb Thanks! That looks to be the fix for my second bullet point, and works as expected after I pull the latest.

@mickeynp Unfortunately that is not what’s happening on my machine. The window layout is being trashed when scrolling with a message view open (this is the situation described in my first bullet point). This used to _also_ mean that after marking, the layout would be trashed as well, but the fix to https://github.com/djcb/mu/issues/2385 has removed auto-view after marking so it’s no longer an issue after marking.

bjcohen commented:
Anyway, I think this fixes the issue: https://github.com/djcb/mu/pull/2395
